src/lsp/ProjectManager.spec.ts

```diff
@@ -341,6 +341,183 @@ describe('ProjectManager', () => {
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/level3/bsconfig.json`, '');
+
+            await manager.syncProjects([{
+                ...workspaceSettings,
+                languageServer: {
+                    ...workspaceSettings.languageServer,
+                    maxDepth: 1
+                }
+            }]);
+
+            // maxDepth: 1 should find files at depth 0 only
+            expect(
+                manager.projects.map(x => x.projectKey).sort()
+            ).to.eql([
+                s`${rootDir}/bsconfig.json`
+            ]);
+        });
+
+        it('respects maxDepth of 5 when discovering projects', async () => {
+            // Create bsconfig.json files at different depths
+            fsExtra.outputFileSync(`${rootDir}/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/level3/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/level3/level4/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/level3/level4/level5/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/level3/level4/level5/level6/bsconfig.json`, '');
+
+            await manager.syncProjects([{
+                ...workspaceSettings,
+                languageServer: {
+                    ...workspaceSettings.languageServer,
+                    maxDepth: 5
+                }
+            }]);
+
+            // maxDepth: 5 should find files at depths 0, 1, 2, 3, 4
+            expect(
+                manager.projects.map(x => x.projectKey).sort()
+            ).to.eql([
+                s`${rootDir}/bsconfig.json`,
+                s`${rootDir}/level1/bsconfig.json`,
+                s`${rootDir}/level1/level2/bsconfig.json`,
+                s`${rootDir}/level1/level2/level3/bsconfig.json`,
+                s`${rootDir}/level1/level2/level3/level4/bsconfig.json`
+            ]);
+        });
+
+        it('respects maxDepth of 20 when discovering projects', async () => {
+            // Create bsconfig.json files at different depths
+            fsExtra.outputFileSync(`${rootDir}/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/level3/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/level3/level4/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/level3/level4/level5/bsconfig.json`, '');
+            fsExtra.outputFileSync(`${rootDir}/level1/level2/level3/level4/level5/level6/bsconfig.json`, '');
```

Comment:
Prove that it stops at level 20. Ensure that a file at 20 is found, and 21 is not found. But you don't need to make all 20 files....you can skip some steps in between.